### ISSUE TITLE
linear_cross_entropy: chunked fast path is gated on is_cuda rather than the capabilities it requires

### ISSUE THREAD

**Issue by dwoai22** [pytorch/pytorch#195832]
### Problem

`torch.nn.functional.linear_cross_entropy`'s chunked implementation has a fast path that avoids several scratch allocations by letting `mm` upcast internally via `out_dtype=`. Whether that path is taken is decided by `is_cuda` (`torch/nn/modules/linear_cross_entropy.py:486` on main), which is consulted by eleven downstream flags.

XPU implements every overload this path uses, but cannot reach it:

| Operator | native_functions.yaml | Dispatch |
|---|---|---|
| `mm.dtype_out` | 4164 | `CUDA: _mm_dtype_out_cuda`, `XPU: _mm_dtype_out_xpu` |
| `addmm.dtype_out` | 7092 | `CUDA`, `XPU` |
| `bmm.dtype_out` | 1387 | `CUDA`, `XPU` |
| `baddbmm.dtype_out` | 1066 | `CUDA`, `XPU` |

On XPU the module instead allocates `logits_acc_buf` and `input_chunk_acc_buf` and performs explicit casts it does not need. Since the purpose of this module is to avoid materializing the `(num_batches, num_classes)` logits tensor for large vocabularies, the unnecessary scratch is a direct cost to the case it exists to serve.

### Root cause

`is_cuda` is not wrong - CUDA is the only backend that satisfies both properties it stands for - but it bundles two independent capabilities under a name that mentions neither:

| Capability | Question | Uses |
|---|---|---|
| `mm(out_dtype=)` support | Can `mm` write a destination whose dtype differs from its operands? | 10 |
| fp32 GEMM accumulator | Does a same-dtype `addmm_` keep an fp32 accumulator internally? | 1 |

The second gates skipping a `(num_classes, in_features)` weight-grad scratch and is only observable at `grad_linear_weight.addmm_(logits_downcast.T, input_chunk)`, the single bf16 x bf16 -> bf16 GEMM in the file. Because the two travel together under one device name, a backend cannot satisfy or declare one without the other, and nothing in the file states what a new device would need.

The file's own comment acknowledges the situation as temporary:

```python
# CUDA gates the out_dtype= mm fast path; the non-CUDA path
# (CPU, MPS, XPU, ...) routes mixed-dtype mm through explicit
# casts until out_dtype= is validated on those backends.
```

### A second blocker on the same path

Enabling XPU needs one more change, in a different function. `LinearCrossEntropyOptions._resolve_auto_acc_dtype` (`torch/nn/modules/linear_cross_entropy_options.py:216`) returns fp32 only for CUDA and CPU. On XPU `acc_dtype` therefore never resolves to fp32 under `acc_policy="auto"`, `use_acc_dtype` stays `False`, and every gate above evaluates `False` regardless. Reaching the path today requires an explicit `acc_dtype=torch.float32`.

That check is a third, distinct capability - "does this device have hardware fp16/bf16 support" - and is dtype-dependent (CUDA gates bf16 at SM 8.0 and fp16 at SM 7.0). Both `torch.cuda` and `torch.xpu` expose `is_bf16_supported()`, though there is no fp16 counterpart and neither `torch.cpu` nor `torch.mps` provides either.

### Proposed fix

Split `is_cuda` into two named capability helpers, each with its own list of validated device types and a docstring stating what to verify before adding one. Behavior-preserving on its own; enabling XPU would then be a separate change with its own numerics validation, alongside the `_resolve_auto_acc_dtype` counterpart above. So, proposed solution does not add a new backend, but creates a structure that can be extended with multiple backends after they are validated.

Happy to send a PR if this is something the maintainers want.

### Versions

Source-structure issue, reproducible by reading the files. Observed against main at `3b4bcd4f227` (2026-08-20), `2.15.0a0+git3b4bcd4`.

@albanD @jbschlosser @malfet @mikaylagawarecki

---

This issue was drafted with the assistance of Claude Code and reviewed before posting.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @gujinghui @EikanWang @fengyuan14 @guangyey

**Comment by abhinav-phi** [comment]
Hi! I'd like to work on this.

Plan: I'll investigate the reported behavior, implement the change/fix per the issue description, and add tests to cover it.

Could you please assign this issue to me? Thanks!

/assign take

**Comment by NaOHCC** [comment]
Hi @dwoai22, a PR would be very welcome! I’d be happy to help review and validate it on XPU.
